File: client/lib/definitions/order.ts
import { OrderStatus } from './common';

export type OrderModel = {
  id: string;
  total: number;
  status: OrderStatus;
  customer_name: string;
  customer_email: string;
  payment_info: PaymentInfo;
  shipping_info: ShippingInfoModel;
  products: OrderItemMode[];
  created_at: string;
};

type OrderAttributeSnapshot = {
  name: string;
  value: string;
};
export type OrderItemMode = {
  id: string;
  name: string;
  image_url: string;
  attribute_snapshot: OrderAttributeSnapshot[];
  line_total: number;
  quantity: number;
};

export type ShippingInfoModel = {
  name: string;
  phone: string;
  address: string;
  city: string;
  district: string;
  ward: string;
};

export type PaymentInfo = {
  method: string;
  status: string;
  transaction_id: string;
  amount: number;
};

File: internal/dto/order.go
package dto

import (
	"time"

	"github.com/google/uuid"
	"github.com/thanhphuocnguyen/go-eshop/internal/db/repository"
)

type LineItem struct {
	ID                 string                             `json:"id"`
	VariantID          string                             `json:"variantId"`
	Name               string                             `json:"name"`
	ImageUrl           *string                            `json:"imageUrl"`
	Price              float64                            `json:"price"`
	LineTotal          float64                            `json:"lineTotal"`
	DiscountAmount     float64                            `json:"discountAmount"`
	Quantity           int16                              `json:"quantity"`
	AttributesSnapshot []repository.AttributeDataSnapshot `json:"attributesSnapshot"`
	CreatedAt          time.Time                          `json:"createdAt"`
	UpdatedAt          time.Time                          `json:"updatedAt"`
}

type OrderDiscount struct {
	ID            string  `json:"id"`
	Code          string  `json:"code"`
	Description   *string `json:"description"`
	DiscountType  string  `json:"discountType"`
	DiscountValue float64 `json:"discountValue"`
}

type OrderDetail struct {
	ID            uuid.UUID                          `json:"id"`
	Total         float64                            `json:"total"`
	Status        repository.OrderStatus             `json:"status"`
	CustomerName  string                             `json:"customerName"`
	CustomerEmail string                             `json:"customerEmail"`
	PaymentInfo   PaymentInfo                        `json:"paymentInfo"`
	ShippingInfo  repository.ShippingAddressSnapshot `json:"shippingInfo"`
	LineItems     []LineItem                         `json:"lineItems"`
	Discounts     []OrderDiscount                    `json:"discounts"`
	CreatedAt     time.Time                          `json:"createdAt"`
	UpdatedAt     time.Time                          `json:"updatedAt"`
}

type OrderListItem struct {
	ID            uuid.UUID                `json:"id"`
	Total         float64                  `json:"total"`
	TotalItems    int32                    `json:"totalItems"`
	Status        repository.OrderStatus   `json:"status"`
	PaymentStatus repository.PaymentStatus `json:"paymentStatus"`
	CustomerName  string                   `json:"customerName"`
	CustomerEmail string                   `json:"customerEmail"`
	CreatedAt     time.Time                `json:"createdAt"`
	UpdatedAt     time.Time                `json:"updatedAt"`
}

type OrderItemAttribute struct {
	Name  string `json:"name"`
	Value string `json:"value"`
}
